File: code/code_for_model_selection/code_for_cpp(Brainstorming)/TestMatrixBasedModelSelector.cpp

#include "TestMatrixBasedModelSelector.h"
#include <iostream>
#include <fstream>
#include <cmath>
#include <random>
#include <algorithm>

TestMatrixBasedModelSelector::TestMatrixBasedModelSelector(int n, int d, double delta, int B, int H)
    : n(n), d(d), delta(delta), B(B), H(H) {}

std::vector<std::vector<int>> TestMatrixBasedModelSelector::generate_rrsd_matrix() {
    double p = std::exp(-1.0 / d);
    int r = static_cast<int>((1 - p) * (n - d + 1));
    int m = static_cast<int>(std::exp(1) * d * std::log(2 * n / delta));
    std::vector<std::vector<int>> matrix(m, std::vector<int>(n, 0));
    
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, n - 1);

    for (int i = 0; i < m; ++i) {
        std::set<int> random_columns;
        while (random_columns.size() < r) {
            random_columns.insert(dis(gen));
        }
        for (int col : random_columns) {
            matrix[i][col] = 1;
        }
    }
    return matrix;
}

std::set<int> TestMatrixBasedModelSelector::decode(const std::vector<std::vector<int>>& matrix, const std::set<int>& defective_set) {
    std::set<int> X;
    for (int i = 0; i < matrix[0].size(); ++i) {
        X.insert(i);
    }

    for (const auto& row : matrix) {
        bool all_zero = true;
        for (int i : defective_set) {
            if (row[i] != 0) {
                all_zero = false;
                break;
            }
        }
        if (all_zero) {
            for (int i = 0; i < row.size(); ++i) {
                if (row[i] == 1) {
                    X.erase(i);
                }
            }
        }
    }
    return X;
}

std::pair<std::vector<std::vector<int>>, double> TestMatrixBasedModelSelector::find_best_model() {
    std::vector<std::vector<int>> best_matrix;
    double best_score = -1.0;
    
    for (int b = 0; b < B; ++b) {
        auto matrix = generate_rrsd_matrix();
        double total_score = 0.0;
        
        for (int h = 0; h < H; ++h) {
            std::set<int> defective_set;
            while (defective_set.size() < d) {
                defective_set.insert(rand() % n);
            }
            auto decoded_set = decode(matrix, defective_set);
            int false_positives = std::distance(std::set_difference(decoded_set.begin(), decoded_set.end(),
                                                         defective_set.begin(), defective_set.end(),
                                                         std::inserter(std::set<int>(), std::set<int>().begin())), decoded_set.end());
            total_score += (1.0 - static_cast<double>(false_positives) / n);
        }
        
        double average_score = total_score / H;
        if (average_score > best_score) {
            best_score = average_score;
            best_matrix = matrix;
        }
    }

    return {best_matrix, best_score};
}

int main() {
    TestMatrixBasedModelSelector selector(10, 2, 0.1, 10, 10);
    auto [best_matrix, best_score] = selector.find_best_model();

    // Output the best matrix and score for demonstration
    std::ofstream matrix_file("best_rrsd_matrix.txt");
    for (const auto& row : best_matrix) {
        for (int val : row) {
            matrix_file << val << " ";
        }
        matrix_file << "\n";
    }
    matrix_file.close();

    std::ofstream score_file("best_score.txt");
    score_file << best_score;
    score_file.close();

    return 0;
}

            total_score += (1.0 - static_cast<double>(false_positives) / n);
        }
        
        double average_score = total_score / H;
        if (average_score > best_score) {
            best_score = average_score;
            best_matrix = matrix;
        }
    }

    return {best_matrix, best_score};
}

int main() {
    TestMatrixBasedModelSelector selector(10, 2, 0.1, 10, 10);
    auto [best_matrix, best_score] = selector.find_best_model();

    // Output the best matrix and score for demonstration
    std::ofstream matrix_file("best_rrsd_matrix.txt");
    for (const auto& row : best_matrix) {
        for (int val : row) {
            matrix_file << val << " ";
        }
        matrix_file << "\n";
    }
    matrix_file.close();

    std::ofstream score_file("best_score.txt");
    score_file << best_score;
    score_file.close();

    return 0;
}

File: code/code_for_model_selection/TestMatrixBasedModelSelector.py

import numpy as np
import random

class TestMatrixBasedModelSelector:
    def __init__(self, n, d, delta, B=100, H=100):
        self.n = n
        self.d = d
        self.delta = delta
        self.B = B
        self.H = H

    def generate_rrsd_matrix(self):
        p = np.exp(-1 / self.d)
        r = int((1 - p) * (self.n - self.d + 1))
        m = int(np.exp(1) * self.d * np.log(2 * self.n / self.delta))
        matrix = np.zeros((m, self.n))
        for i in range(m):
            random_columns = np.random.choice(self.n, r, replace=False)
            matrix[i, random_columns] = 1
        return matrix

    def generate_defective_sets(self):
        defective_sets = []
        for _ in range(self.H):
            defective_set = random.sample(range(self.n), self.d)
            defective_sets.append(defective_set)
        return defective_sets

    def decode(self, matrix, defective_set):
        X = set(range(self.n))
        for row in matrix:
            if all(row[i] == 0 for i in defective_set):
                X -= {i for i, val in enumerate(row) if val == 1}
        return X

    def calculate_false_positives(self, matrix, defective_set):
        X = self.decode(matrix, defective_set)
        false_positives = len(X - set(defective_set))
        return false_positives

    def calculate_score(self, false_positives):
        return (1 - false_positives / self.n)

    def find_best_model(self):
        best_matrix = None
        best_score = -1
        defective_sets = self.generate_defective_sets()
        
        for b in range(self.B):
            matrix = self.generate_rrsd_matrix()
            total_score = 0
            correct_decodes = 0
            print(f"For Test Matrix {b + 1}:")
            
            for defective_set in defective_sets:
                false_positives = self.calculate_false_positives(matrix, defective_set)
                decoded_set = self.decode(matrix, defective_set)
                
                if set(defective_set) == decoded_set:
                    correct_decodes += 1

                #print(f"Defective Set: {defective_set}")
                #print(f"Decoded Set: {decoded_set}")
                
                total_score += self.calculate_score(false_positives)
            
            average_score = total_score / self.H
            print(f"Correctly Decoded: {correct_decodes} out of {self.H}")
            print(f"Current score:{average_score}")
            
            if average_score > best_score:
                best_score = average_score
                best_matrix = matrix
                
        return {'best_matrix': best_matrix, 'best_score': best_score}

# Example usage
if __name__ == "__main__":
    selector = TestMatrixBasedModelSelector(n=2048, d=16, delta=0.1, B=100, H=100)
    result = selector.find_best_model()
    best_matrix = result['best_matrix']
    best_score = result['best_score']
    print("Best Matrix:")
    print(best_matrix)
    print("Best Score:", best_score)
    print("Best matrix shape {}".format(best_matrix.shape))
